src/ol/layer/WebGLTile.js

```diff
@@ -0,0 +1,330 @@
+ * this layer in its layers collection, and the layer will be rendered on top. This is useful for
+ * temporary layers. The standard way to add a layer to a map and have it managed by the map is to
+ * use {@link module:ol/Map#addLayer}.
+ * @property {boolean} [useInterimTilesOnError=true] Use interim tiles on error.
+ */
+
+/**
+ * @typedef {Object} ParsedStyle
+ * @property {string} vertexShader The vertex shader.
+ * @property {string} fragmentShader The fragment shader.
+ * @property {Object<string,import("../webgl/Helper.js").UniformValue>} uniforms Uniform definitions.
+ */
+
+/**
+ * @param {Style} style The layer style.
+ * @return {ParsedStyle} Shaders and uniforms generated from the style.
+ */
+function parseStyle(style) {
+  const vertexShader = `
+    attribute vec2 ${Attributes.TEXTURE_COORD};
+    uniform mat4 ${Uniforms.TILE_TRANSFORM};
+    uniform float ${Uniforms.DEPTH};
+
+    varying vec2 v_textureCoord;
+
+    void main() {
+      v_textureCoord = ${Attributes.TEXTURE_COORD};
+      gl_Position = ${Uniforms.TILE_TRANSFORM} * vec4(${Attributes.TEXTURE_COORD}, ${Uniforms.DEPTH}, 1.0);
+    }
+  `;
+
+  /**
+   * @type {import("../style/expressions.js").ParsingContext}
+   */
+  const context = {
+    inFragmentShader: true,
+    variables: [],
+    attributes: [],
+    stringLiteralsMap: {},
+  };
+
+  const pipeline = [];
+
+  if (style.math !== undefined) {
+    const math = expressionToGlsl(context, style.math, ValueTypes.NUMBER);
+    pipeline.push(`color.rgb = vec3(${math});`);
+  }
+
+  if (style.colors !== undefined) {
+    const colorCount = style.colors.length / 2;
+    if (colorCount % 1 !== 0) {
+      throw new Error(
+        'The colors array length must be even (one value per color)'
+      );
+    }
+
+    for (let i = 0; i < colorCount - 1; ++i) {
+      const offset = i * 2;
+
+      const startValue = expressionToGlsl(
+        context,
+        style.colors[offset],
+        ValueTypes.NUMBER
+      );
+      const startColor = expressionToGlsl(
+        context,
+        style.colors[offset + 1],
+        ValueTypes.COLOR
+      );
+
+      if (i === 0) {
+        pipeline.push(`
+          if (color.r < ${startValue}) {
+            color = ${startColor};
+          }
+        `);
+      }
+
+      const endValue = expressionToGlsl(
+        context,
+        style.colors[offset + 2],
+        ValueTypes.NUMBER
+      );
+      const endColor = expressionToGlsl(
+        context,
+        style.colors[offset + 3],
+        ValueTypes.COLOR
+      );
+
+      pipeline.push(`
+        else if (color.r < ${endValue}) {
+          color = mix(${startColor}, ${endColor}, (color.r - ${startValue}) / (${endValue} - ${startValue}));
+        }
+      `);
+
+      if (i + 2 === colorCount) {
+        pipeline.push(`
+          else if (color.r >= ${endValue}) {
+            color = ${endColor};
+          }
+        `);
+      }
+    }
+  }
```

Comment:
A significant difference is that these are "statements" instead of just "expressions."  I'm going to leave them here for now, since they are specific to WebGLTile rendering.  If we find a case for sharing the code with other components later, we can shuffle it around.